charmhelpers/contrib/openstack/vaultlocker.py

```diff
@@ -52,29 +71,70 @@ def __call__(self):
+                        if "sentinel" in current_tokens:
+                            current_tokens = [token]
+                        else:
+                            current_tokens.append(token)
+
+                    # If we have already had a successful request then all we
+                    # are doing is collecting current tokens so that we can
+                    # purge old ones.
+                    if self.complete:
+                        continue
+
                     # Tokens may change when secret_id's are being
                     # reissued - if so use token to get new secret_id
-                    if token != last_token:
-                        secret_id = retrieve_secret_id(
-                            url=vault_url,
-                            token=token
-                        )
+                    success = False
+                    if token not in used_tokens:
+                        try:
+                            secret_id = retrieve_secret_id(
+                                url=vault_url,
+                                token=token
+                            )
+                            success = True
+                        except hvac.exceptions.InvalidRequest:
+                            # Try next
+                            pass
+
+                    if success:
+                        used_tokens[token] = secret_id
                         db.set('secret-id', secret_id)
                         db.set('last-token', token)
                         db.flush()
 
-                    ctxt = {
-                        'vault_url': vault_url,
-                        'role_id': json.loads(role_id),
-                        'secret_id': secret_id,
-                        'secret_backend': self.secret_backend,
-                    }
-                    vault_ca = data.get('vault_ca')
-                    if vault_ca:
-                        ctxt['vault_ca'] = json.loads(vault_ca)
-                    self.complete = True
-                    return ctxt
-        return {}
+                        ctxt['vault_url'] = vault_url
+                        ctxt['role_id'] = json.loads(role_id)
+                        ctxt['secret_id'] = secret_id
+                        ctxt['secret_backend'] = self.secret_backend
+
+                        vault_ca = data.get('vault_ca')
+                        if vault_ca:
+                            ctxt['vault_ca'] = json.loads(vault_ca)
+
+                        self.complete = True
+                    else:
+                        if token not in used_tokens:
+                            used_tokens[token] = None
+                        elif used_tokens[token] == secret_id:
+                            ctxt['vault_url'] = vault_url
+                            ctxt['role_id'] = json.loads(role_id)
+                            ctxt['secret_id'] = secret_id
+                            ctxt['secret_backend'] = self.secret_backend
```

Comment:
Should probably also handle changes to the `vault_ca` in this case.

Reply:
ack ill fix that